src/gui/PdfFloatingToolbox.cpp

```diff
@@ -0,0 +1,188 @@
+    view->pdfTextSelection = nullptr;
+
+    view->rerenderPage();
+}
+
+void PdfFloatingToolbox::highlightCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->createStrokesForHighlight();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::copyTextCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->copyText();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::underlineCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->createStrokesForUnderline();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::strikethroughCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->createStrokesForStrikethrough();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::closeCb(GtkButton* button, PdfFloatingToolbox* pft) { pft->postAction(); }
+
+void PdfFloatingToolbox::show() {
+    gtk_widget_hide(this->floatingToolbox);  // force showing in new position
+    gtk_widget_show_all(this->floatingToolbox);
+}
+
+void PdfFloatingToolbox::setSelectType(PdfTextSelectType type) { this->selectType = type; }
+
+PdfTextSelectType PdfFloatingToolbox::getSelectType() { return this->selectType; }
+
+void PdfFloatingToolbox::switchSelectTypeCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->switchSelectType();
+    pft->pdfTextSelection->select();
+    pft->pdfTextSelection->getPageView()->rerenderPage();
+}
+
+void PdfFloatingToolbox::switchSelectType() {
+    if (this->selectType != PdfTextSelectType::SELECT_IN_AREA) {
+        this->selectType = PdfTextSelectType::SELECT_IN_AREA;
+    } else {
+        this->selectType = PdfTextSelectType::SELECT_HEAD_TAIL;
+    }
+}
+
+void PdfFloatingToolbox::copyText() {
+    GtkClipboard* clipboard = gtk_widget_get_clipboard(this->theMainWindow->getWindow(), GDK_SELECTION_CLIPBOARD);
+
+    const gchar* text = this->pdfTextSelection->getSelectedText().c_str();
+
+    gtk_clipboard_set_text(clipboard, text, -1);
+}
+
+void PdfFloatingToolbox::createStrokes(PdfMarkerStyle position, PdfMarkerStyle width, int markerOpacity) {
+    if (this->pdfTextSelection->getSelectedTextRects().empty())
+        return;
+
+    auto strokeHandler =
+            new StrokeHandler(this->pdfTextSelection->getPageView()->getXournal(),
+                              this->pdfTextSelection->getPageView(), this->pdfTextSelection->getPageView()->getPage());
+    Stroke* stroke = nullptr;
+
+    auto p = new PositionInputData();
+    p->pressure = -1;
+    p->state = GDK_RELEASE_MASK;
+
+    for (XojPdfRectangle rect: this->pdfTextSelection->getSelectedTextRects()) {
+        // the center line position of stroke
+        double h = position == PdfMarkerStyle::POS_TEXT_BOTTOM ? std::max(rect.y1, rect.y2) :
+                   position == PdfMarkerStyle::POS_TEXT_MIDDLE ? (rect.y1 + rect.y2) / 2 :
+                                                                 std::min(rect.y1, rect.y2);
+
+        // the width of stroke
+        double w = width == PdfMarkerStyle::WIDTH_TEXT_LINE ? 1 : std::abs(rect.y2 - rect.y1);
```

Comment:
I think I find the following might be easier to understand:

```suggestion
    for (XojPdfRectangle rect: this->pdfTextSelection->getSelectedTextRects()) {
        double topOfLine = std::min(rect.y1, rect.y2);
        double middleOfLine = (rect.y1 + rect.y2) / 2;
        double bottomOfLine = std::max(rect.y1, rect.y2);
        double rectWidth = std::abs(rect.y2 - rect.y1);
        
        // the center line position of stroke
        double h = position == PdfMarkerStyle::POS_TEXT_BOTTOM ? bottomOfLine :
                   position == PdfMarkerStyle::POS_TEXT_MIDDLE ? middleOfLine : topOfLine;

        // the width of stroke
        double w = width == PdfMarkerStyle::WIDTH_TEXT_LINE ? 1 : rectWidth;
```

This is subjective! If you disagree with this change, please ignore it!

Reply:
We can clean this up further by making `PdfMarkerStyle` an `enum class` and then using a `switch` statement. The compiler will warn about missing cases, which can potentially catch some bugs.